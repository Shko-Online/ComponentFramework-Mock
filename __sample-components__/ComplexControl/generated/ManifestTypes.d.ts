/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    inputProperty: ComponentFramework.PropertyTypes.TwoOptionsProperty;
    boundProperty: ComponentFramework.PropertyTypes.TwoOptionsProperty;
}
export interface IOutputs {
    boundProperty?: boolean;
    outputProperty?: any;
}
