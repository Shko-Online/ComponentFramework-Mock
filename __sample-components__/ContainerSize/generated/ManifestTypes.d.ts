/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    trackContainer: ComponentFramework.PropertyTypes.TwoOptionsProperty;
}
export interface IOutputs {
    trackContainer?: boolean;
}
