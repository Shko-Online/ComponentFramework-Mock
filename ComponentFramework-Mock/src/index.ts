/*
Copyright (c) 2022 Betim Beja and Shko Online LLC
Licensed under the MIT license.
*/

/// <reference types="powerapps-component-framework" />
/// <reference path="./global.d.ts" />

export type { MockToRaw, PropertyMap, PropertyToMock } from './ComponentFramework-Mock';
export type { MockGenerator } from './ComponentFramework-Mock-Generator';
export type { ShkoOnline } from './ShkoOnline';

export {
    AttributeMetadataCollection,
    AttributeType,
    ClientMock,
    ContextMock,
    DataSetMock,
    DateTimeMetadataMock,
    DateTimePropertyMock,
    DecimalNumberMetadataMock,
    DecimalNumberPropertyMock,
    DeviceMock,
    EntityRecordMock,
    EnumPropertyMock,
    FactoryMock,
    FilteringMock,
    FormattingMock,
    LinkingMock,
    LookupMetadataMock,
    LookupPropertyMock,
    MetadataMock,
    ModeMock,
    MultiSelectOptionSetPropertyMock,
    NavigationMock,
    NumberMetadataMock,
    NumberPropertyMock,
    OptionMetadataMock,
    OptionSetMetadataMock,
    OptionSetPropertyMock,
    PagingMock,
    PopupServiceMock,
    PropertyMock,
    ResourcesMock,
    StringMetadataMock,
    StringPropertyMock,
    TwoOptionMetadataMock,
    TwoOptionsPropertyMock,
    UserSettingsMock,
    UtilityMock,
    WebApiMock,
    WholeNumberMetadataMock,
    WholeNumberPropertyMock,
} from './ComponentFramework-Mock';

export {
    ComponentFrameworkMockGeneratorReact,
    ComponentFrameworkMockGenerator,
    MetadataDB,
    ReactResizeObserver,
    mockGetEntityMetadata,
    mockRefreshParameters,
    mockSetControlResource,
    mockSetControlState,
} from './ComponentFramework-Mock-Generator';

export { AttributeMetadataGenerator, arrayEqual } from './utils';
