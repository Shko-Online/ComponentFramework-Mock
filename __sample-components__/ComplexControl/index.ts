import type { IInputs, IOutputs } from './generated/ManifestTypes';

export class ComplexControl implements ComponentFramework.StandardControl<IInputs, IOutputs> {
    private boundProperty: boolean;
    private inputProperty: boolean;
    private notifyOutputChanged: () => void;
    private updateCount = 0;
    private container: HTMLDivElement;
    /**
     * Empty constructor.
     */
    constructor() {}

    public async getOutputSchema(context: ComponentFramework.Context<IInputs>): Promise<Record<string, unknown>> {
        return Promise.resolve({
            outputProperty: {
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'object',
                properties: {
                    inputProperty: {
                        type: 'boolean',
                    },
                    updateCount: {
                        type: 'number',
                    },
                },
            },
        });
    }

    /**
     * Used to initialize the control instance. Controls can kick off remote server calls and other initialization actions here.
     * Data-set values are not initialized here, use updateView.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to property names defined in the manifest, as well as utility functions.
     * @param notifyOutputChanged A callback method to alert the framework that the control has new outputs ready to be retrieved asynchronously.
     * @param state A piece of data that persists in one session for a single user. Can be set at any point in a controls life cycle by calling 'setControlState' in the Mode interface.
     * @param container If a control is marked control-type='standard', it will receive an empty div element within which it can render its content.
     */
    public init(
        context: ComponentFramework.Context<IInputs>,
        notifyOutputChanged: () => void,
        state: ComponentFramework.Dictionary,
        container: HTMLDivElement,
    ): void {
        this.notifyOutputChanged = notifyOutputChanged;
        // Add control initialization code
        this.container = container;
        this.container.innerHTML = `Updated ${this.updateCount} ${this.updateCount === 1 ? 'time' : 'times'}`;
        this.inputProperty = context.parameters.inputProperty.raw;
        this.boundProperty = context.parameters.boundProperty.raw;
    }

    /**
     * Called when any value in the property bag has changed. This includes field values, data-sets, global values such as container height and width, offline status, control metadata values such as label, visible, etc.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to names defined in the manifest, as well as utility functions
     */
    public updateView(context: ComponentFramework.Context<IInputs>): void {
        // Add control initialization code
        this.updateCount += 1;
        this.container.innerHTML = `Updated ${this.updateCount} ${this.updateCount === 1 ? 'time' : 'times'}`;
        this.boundProperty = context.parameters.boundProperty.raw;
        if (this.inputProperty !== context.parameters.inputProperty.raw) {
            this.inputProperty = context.parameters.inputProperty.raw;
            this.notifyOutputChanged();
        }
    }
    
    /**
     * It is called by the framework prior to a control receiving new data.
     * @returns an object based on nomenclature defined in manifest, expecting object[s] for property marked as “bound” or “output”
     */
    public getOutputs(): IOutputs {
        const updateCount = this.updateCount;
        const boundProperty = this.boundProperty;
        const inputProperty = this.inputProperty;
        const outputProperty = { updateCount, inputProperty };
        return { boundProperty, outputProperty };
    }

    /**
     * Called when the control is to be removed from the DOM tree. Controls should use this call for cleanup.
     * i.e. cancelling any pending remote calls, removing listeners, etc.
     */
    public destroy(): void {
        // Add code to cleanup control if necessary
    }
}
