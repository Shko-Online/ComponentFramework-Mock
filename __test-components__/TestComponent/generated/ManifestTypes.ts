/*
*This is auto generated from the ControlManifest.Input.xml file
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    selection: ComponentFramework.PropertyTypes.MultiSelectOptionSetProperty;
    controlType: ComponentFramework.PropertyTypes.EnumProperty<"0" | "1">;
    orientation: ComponentFramework.PropertyTypes.EnumProperty<"1" | "0">;
}
export interface IOutputs {
    selection?: number[];
}
