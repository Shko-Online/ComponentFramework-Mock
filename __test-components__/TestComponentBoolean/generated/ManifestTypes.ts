/*
*This is auto generated from the ControlManifest.Input.xml file
*/

// Define IInputs and IOutputs Type. They should match with ControlManifest.
export interface IInputs {
    turnedOn: ComponentFramework.PropertyTypes.TwoOptionsProperty;
}
export interface IOutputs {
    turnedOn?: boolean;
}
