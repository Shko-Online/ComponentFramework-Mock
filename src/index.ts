/*
Copyright (c) 2022 Betim Beja and Shko Online LLC
Licensed under the MIT license.
*/

export type { MockToRaw, PropertyMap, PropertyToMock } from './ComponentFramework-Mock';
export type { MockGenerator } from './ComponentFramework-Mock-Generator';
export type { ShkoOnline } from './ShkoOnline';

export {
    AttributeMetadataCollection,
    AttributeType,
    ClientMock,
    ContextMock,
    DataSetMock,
    DateTimeMetadataMock,
    DateTimePropertyMock,
    DecimalNumberMetadataMock,
    DecimalNumberPropertyMock,
    DeviceMock,
    EntityRecordMock,
    EnumPropertyMock,
    FactoryMock,
    FilteringMock,
    FormattingMock,
    LinkingMock,
    LookupMetadataMock,
    LookupPropertyMock,
    MetadataMock,
    ModeMock,
    MultiSelectOptionSetPropertyMock,
    NavigationMock,
    NumberMetadataMock,
    NumberPropertyMock,
    OptionMetadataMock,
    OptionSetMetadataMock,
    OptionSetPropertyMock,
    OptionSetType,
    PagingMock,
    PopupServiceMock,
    PropertyMock,
    ResourcesMock,
    StringMetadataMock,
    StringPropertyMock,
    TwoOptionMetadataMock,
    TwoOptionsPropertyMock,
    UserSettingsMock,
    UtilityMock,
    WebApiMock,
    WholeNumberMetadataMock,
    WholeNumberPropertyMock,
} from './ComponentFramework-Mock';

export {
    ComponentFrameworkMockGeneratorReact,
    ComponentFrameworkMockGenerator,
    MetadataDB,
    ReactResizeObserver,
    mockGetEntityMetadata,
    mockRefreshDatasets,
    mockRefreshParameters,
    mockSetControlResource,
    mockSetControlState,
} from './ComponentFramework-Mock-Generator';

export { arrayEqual, AttributeMetadataGenerator, itemEqual, newGuid, showBanner } from './utils';
