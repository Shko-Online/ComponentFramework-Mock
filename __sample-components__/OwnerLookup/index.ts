/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import { IInputs, IOutputs } from './generated/ManifestTypes';

export class OwnerLookup implements ComponentFramework.StandardControl<IInputs, IOutputs> {
    private container: HTMLDivElement;
    private notifyOutputChanged: () => void;
    private myself: ComponentFramework.LookupValue[];
    private value?: ComponentFramework.LookupValue[];
    /**
     * Empty constructor.
     */
    constructor() {}

    /**
     * Used to initialize the control instance. Controls can kick off remote server calls and other initialization actions here.
     * Data-set values are not initialized here, use updateView.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to property names defined in the manifest, as well as utility functions.
     * @param notifyOutputChanged A callback method to alert the framework that the control has new outputs ready to be retrieved asynchronously.
     * @param state A piece of data that persists in one session for a single user. Can be set at any point in a controls life cycle by calling 'setControlState' in the Mode interface.
     * @param container If a control is marked control-type='standard', it will receive an empty div element within which it can render its content.
     */
    public init(
        context: ComponentFramework.Context<IInputs>,
        notifyOutputChanged: () => void,
        state: ComponentFramework.Dictionary,
        container: HTMLDivElement,
    ): void {
        // Add control initialization code
        this.container = container;
        this.myself = [{ entityType: 'systemuser', id: context.userSettings.userId, name: 'Assigned from PCF' }];
        const name = document.createElement('p');
        this.updateName(context, name);
        container.appendChild(name);

        const button = document.createElement('button');
        button.innerHTML = 'Assign To Myself';
        button.addEventListener('click', this.assignToMyself.bind(this), false);
        container.appendChild(button);
        this.notifyOutputChanged = notifyOutputChanged;
    }

    /**
     * Called when any value in the property bag has changed. This includes field values, data-sets, global values such as container height and width, offline status, control metadata values such as label, visible, etc.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to names defined in the manifest, as well as utility functions
     */
    public updateView(context: ComponentFramework.Context<IInputs>): void {
        // Add code to update control view
        const name = this.container.getElementsByTagName('p')[0];
        this.updateName(context, name);
    }

    private updateName(context: ComponentFramework.Context<IInputs>, name: HTMLParagraphElement) {
        const value = context.parameters.value.raw;
        this.value = value;
        name.innerHTML = value && value.length ? `[${value[0].entityType}-${value[0].id}] ${value[0].name}` : '---';
    }

    private assignToMyself() {
        this.value = this.myself;
        this.notifyOutputChanged();
    }

    /**
     * It is called by the framework prior to a control receiving new data.
     * @returns an object based on nomenclature defined in manifest, expecting object[s] for property marked as “bound” or “output”
     */
    public getOutputs(): IOutputs {
        const { value } = this;
        return { value };
    }

    /**
     * Called when the control is to be removed from the DOM tree. Controls should use this call for cleanup.
     * i.e. cancelling any pending remote calls, removing listeners, etc.
     */
    public destroy(): void {
        // Add code to cleanup control if necessary
    }
}
