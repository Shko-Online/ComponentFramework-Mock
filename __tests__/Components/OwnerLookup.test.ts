/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import { it, expect, describe, beforeEach } from '@jest/globals';
import { ComponentFrameworkMockGenerator, LookupPropertyMock } from '../../src';
import { OwnerLookup } from '../../__sample-components__/OwnerLookup';
import { IInputs, IOutputs } from '../../__sample-components__/OwnerLookup/generated/ManifestTypes';

describe('OwnerLookup', () => {
    let mockGenerator: ComponentFrameworkMockGenerator<IInputs, IOutputs>;
    beforeEach(() => {
        const container = document.createElement('div');
        mockGenerator = new ComponentFrameworkMockGenerator(
            OwnerLookup,
            {
                value: LookupPropertyMock,
            },
            container,
        );
        mockGenerator.context._SetCanvasItems({
            value: {
                entityType: 'team',
                id: 'guid1',
                name: 'Shko Online',
            },
        });
        mockGenerator.context.userSettings.userId = 'MyUserId';
        mockGenerator.context.userSettings.userName = 'Betim Beja';
        document.body.appendChild(container);
    });

    afterEach(() => {
        document.body.innerHTML = null;
    });

    it('Should render lookup value', () => {
        mockGenerator.ExecuteInit();
        mockGenerator.ExecuteUpdateView();
        expect(mockGenerator.container).toMatchSnapshot();
    });

    it('Should update value on click', () => {
        mockGenerator.ExecuteInit();
        mockGenerator.ExecuteUpdateView();

        const button = mockGenerator.container.getElementsByTagName('button')[0];
        var evt = document.createEvent('Event');
        evt.initEvent('click', false, true);
        button.dispatchEvent(evt);

        expect(mockGenerator.context.updatedProperties).toStrictEqual(['value', 'parameters']);
        expect(document.body).toMatchSnapshot();
    });
});
