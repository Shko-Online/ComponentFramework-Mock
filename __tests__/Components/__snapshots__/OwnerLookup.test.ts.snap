// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnerLookup Should render lookup value 1`] = `
<div>
  <p>
    [team-guid1] Shko Online
  </p>
  <button>
    Assign To Myself
  </button>
</div>
`;

exports[`OwnerLookup Should update value on click 1`] = `
<body>
  <div>
    <p>
      [systemuser-MyUserId] Assigned from PCF
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
</body>
`;
